File: web/dr-viewer/analyze/mixinConfigIndex.js
/**
 * Port of MixinConfigIndex.java — maps mixin config paths/basenames to owning mods (CA-01).
 */

function str(o, key) {
  if (o == null || o[key] == null) return null;
  try {
    return String(o[key]);
  } catch {
    return null;
  }
}

function basename(path) {
  const bang = path.lastIndexOf('!');
  const after = bang >= 0 ? path.slice(bang + 1) : path;
  const slash = after.lastIndexOf('/');
  return slash >= 0 ? after.slice(slash + 1) : after;
}

function put(map, path, hit) {
  const normalized = path.replace(/\\/g, '/');
  if (!map.has(normalized)) map.set(normalized, hit);
  const lower = normalized.toLowerCase();
  if (!map.has(lower)) map.set(lower, hit);
  const base = basename(normalized);
  if (!map.has(base)) map.set(base, hit);
  const baseLower = base.toLowerCase();
  if (!map.has(baseLower)) map.set(baseLower, hit);
}

export function fromMods(mods) {
  const map = new Map();
  if (!Array.isArray(mods)) {
    return createIndex(map);
  }
  for (const mod of mods) {
    if (!mod || typeof mod !== 'object') continue;
    let modId = str(mod, 'id') || str(mod, 'mod_id');
    if (!modId || !modId.trim()) continue;
    let jarName = str(mod, 'jar_file') || str(mod, 'jar');
    if (Array.isArray(mod.mixin_configs)) {
      for (const path of mod.mixin_configs) {
        if (path == null || !String(path).trim()) continue;
        put(map, String(path), { modId, jarName, configPath: String(path) });
      }
    }
    if (Array.isArray(mod.jar_in_jar)) {
      for (const nested of mod.jar_in_jar) {
        if (!nested || typeof nested !== 'object') continue;
        const nestedId = str(nested, 'id') || modId;
        const nestedJar = str(nested, 'jar_file');
        const indexJar = jarName && nestedJar
          ? `${jarName}!${nestedJar}`
          : (nestedJar || jarName);
        if (!Array.isArray(nested.mixin_configs)) continue;
        for (const path of nested.mixin_configs) {
          if (path == null || !String(path).trim()) continue;
          put(map, String(path), { modId: nestedId, jarName: indexJar, configPath: String(path) });
        }
      }
    }
  }
  return createIndex(map);
}

export function empty() {
  return createIndex(new Map());
}

function createIndex(byKey) {
  return {
    resolve(configToken) {
      if (!configToken || !String(configToken).trim()) return null;
      const token = String(configToken).trim().replace(/\\/g, '/');
      let hit = byKey.get(token);
      if (hit) return hit;
      hit = byKey.get(token.toLowerCase());
      if (hit) return hit;
      const base = basename(token);
      hit = byKey.get(base);
      if (hit) return hit;
      return byKey.get(base.toLowerCase()) || null;
    },
    isEmpty() {
      return byKey.size === 0;
    },
  };
}

File: watchtower-core/src/main/java/dev/mcstatus/watchtower/core/collect/MixinConfigIndex.java
package dev.mcstatus.watchtower.core.collect;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;

/**
 * Maps mixin config paths / basenames to owning mods from {@code optional.mods[]} (CA-01 index).
 */
public final class MixinConfigIndex {

    public record Hit(String modId, String jarName, String configPath) {
    }

    private final Map<String, Hit> byKey;

    private MixinConfigIndex(Map<String, Hit> byKey) {
        this.byKey = byKey;
    }

    public static MixinConfigIndex fromMods(JsonArray mods) {
        Map<String, Hit> map = new HashMap<>();
        if (mods == null) {
            return new MixinConfigIndex(map);
        }
        for (JsonElement el : mods) {
            if (!el.isJsonObject()) {
                continue;
            }
            JsonObject mod = el.getAsJsonObject();
            String modId = str(mod, "id");
            if (modId == null || modId.isBlank()) {
                modId = str(mod, "mod_id");
            }
            if (modId == null || modId.isBlank()) {
                continue;
            }
            String jarName = str(mod, "jar_file");
            if (jarName == null) {
                jarName = str(mod, "jar");
            }
            if (mod.has("mixin_configs") && mod.get("mixin_configs").isJsonArray()) {
                for (JsonElement cfgEl : mod.getAsJsonArray("mixin_configs")) {
                    if (!cfgEl.isJsonPrimitive()) {
                        continue;
                    }
                    String path = cfgEl.getAsString();
                    if (path == null || path.isBlank()) {
                        continue;
                    }
                    put(map, path, new Hit(modId, jarName, path));
                }
            }
            // Nested jar-in-jar configs encoded as parent.jar!nested/path
            if (mod.has("jar_in_jar") && mod.get("jar_in_jar").isJsonArray()) {
                for (JsonElement nestedEl : mod.getAsJsonArray("jar_in_jar")) {
                    if (!nestedEl.isJsonObject()) {
                        continue;
                    }
                    JsonObject nested = nestedEl.getAsJsonObject();
                    String nestedId = str(nested, "id");
                    if (nestedId == null) {
                        nestedId = modId;
                    }
                    String nestedJar = str(nested, "jar_file");
                    String indexJar = jarName != null && nestedJar != null
                            ? jarName + "!" + nestedJar
                            : (nestedJar != null ? nestedJar : jarName);
                    if (!nested.has("mixin_configs") || !nested.get("mixin_configs").isJsonArray()) {
                        continue;
                    }
                    for (JsonElement cfgEl : nested.getAsJsonArray("mixin_configs")) {
                        if (!cfgEl.isJsonPrimitive()) {
                            continue;
                        }
                        String path = cfgEl.getAsString();
                        if (path == null || path.isBlank()) {
                            continue;
                        }
                        put(map, path, new Hit(nestedId, indexJar, path));
                    }
                }
            }
        }
        return new MixinConfigIndex(map);
    }

    public static MixinConfigIndex empty() {
        return new MixinConfigIndex(Map.of());
    }

    public Optional<Hit> resolve(String configToken) {
        if (configToken == null || configToken.isBlank()) {
            return Optional.empty();
        }
        String token = configToken.strip().replace('\\', '/');
        Hit hit = byKey.get(token);
        if (hit != null) {
            return Optional.of(hit);
        }
        hit = byKey.get(token.toLowerCase(Locale.ROOT));
        if (hit != null) {
            return Optional.of(hit);
        }
        String base = basename(token);
        hit = byKey.get(base);
        if (hit != null) {
            return Optional.of(hit);
        }
        return Optional.ofNullable(byKey.get(base.toLowerCase(Locale.ROOT)));
    }

    public boolean isEmpty() {
        return byKey.isEmpty();
    }

    private static void put(Map<String, Hit> map, String path, Hit hit) {
        String normalized = path.replace('\\', '/');
        map.putIfAbsent(normalized, hit);
        map.putIfAbsent(normalized.toLowerCase(Locale.ROOT), hit);
        String base = basename(normalized);
        map.putIfAbsent(base, hit);
        map.putIfAbsent(base.toLowerCase(Locale.ROOT), hit);
    }

    private static String basename(String path) {
        int bang = path.lastIndexOf('!');
        String after = bang >= 0 ? path.substring(bang + 1) : path;
        int slash = after.lastIndexOf('/');
        return slash >= 0 ? after.substring(slash + 1) : after;
    }

    private static String str(JsonObject o, String key) {
        if (o == null || !o.has(key) || o.get(key).isJsonNull()) {
            return null;
        }
        try {
            return o.get(key).getAsString();
        } catch (Exception e) {
            return null;
        }
    }
}
